### PR TITLE
Add partial (listing and decompressing only) support for RAR

### PR DESCRIPTION
**Pull request by lmkra**
Attempt to address #152 using https://github.com/muja/unrar.rs crate.

### CODE REVIEW

File: CHANGELOG.md
Comment (marcospb19) [inline CHANGELOG.md:1]: Thanks for the PR!

This change looks unintentional.

CI is failing to compile the `unrar-sys` on `unknown-linux-musl`, does it require installing dependencies with `apt`?

Comment (lmkra) [reply CHANGELOG.md]: Ooops, that definitely was unintentional.

As for `unrar-sys` on musl-based systems, [it may be non-trivial](https://github.com/clux/muslrust/issues/114), but I'll look into it.

Comment (lmkra) [reply CHANGELOG.md]: I've tried to reproduce the build error, but failed to achieve that - linking statically against musl works fine for me.
(disclaimer: I do not use nor have at hand an Ubuntu machine)

I suspect it may be related to the specific setup of github actions environment.
I've pushed a workaround that links system `g++` to `musl-g++` to see whether any further errors will arise during tests.

Comment (lmkra) [reply CHANGELOG.md]: Thank you for running CI again, it helped me to reproduce the error.

Main issue here is that:
1) musl-tools does not provide C++ runtime linked against musl
2) libstdc++ included in github runner environment is linked against glibc

Easiest solution would be not to compile `-musl` builds directly on the runner system, but use cross environment (which includes full C++ toolchain linked against musl).

**Comment by lmkra** [comment]
ARMv7 CI failure seems to be an upstream issue: https://github.com/muja/unrar.rs/issues/35

**Comment by marcospb19** [comment]
Thanks for the follow ups, I temporarly disabled ARMv7 to see if CI succeeds in all other targets.

**Comment by marcospb19** [comment]
Thanks for the follow ups!

I temporarly disabled ARMv7 to see if CI succeeds in other targets, it's failing on Windows, let me know whether you know how to fix it.

**Comment by lmkra** [comment]
MSVC target on Windows should work fine, it's only MinGW one that fails.
I've proposed a fix for that: https://github.com/muja/unrar.rs/pull/37

In the meantime I've temporarily switched used unrar.rs version to my own fork, which contains all of upstream-proposed fixes.

**Comment by lmkra** [comment]
Situation now seems to be that:
* x86_64-pc-windows-gnu target builds correctly
* ...but integrations tests fail due to ouch binary being dynamically linked to libstdc++-6.dll, which is provided by MinGW compiler and is not found in standard Windows shell search path.

While etting up $PATH variable to MinGW install directory or shipping aforementioned library together with ouch.exe could be considered as a workarounds, but I think that better solution would be to statically link against libstdc++ in ouch executable.
Just need to find out how to convince cargo to do it...

**Comment by lmkra** [comment]
Seems to be working now 😃

**Review by marcospb19 — APPROVED** [review]
> Seems to be working now 😃

Amazing!

I think it's better not to merge right now because it would block new `Ouch` releases from happening, I can only publish a version on `crates.io` when all dependencies are fetched from `crates.io` itself, I guess we'll have to wait for a new `unrar` release (but we don't know when that comes out).

Does that make sense to you? :thinking:

**Comment by lmkra** [comment]
Thanks!

And yes, that makes perfect sense to me 👍

**Comment by lmkra** [comment]
New unrar has finally landed on crates.io :)
Therefore I've rebased this PR against current master branch and removed all temporary patching.

**Comment by marcospb19** [comment]
That's awesome! Thank you for keeping an eye on that.

File: Cargo.toml
Comment (marcospb19) [inline Cargo.toml:19]: I did a last re-review, and found this which is unusual, I'm assuming it happened when you solved some merge conflict.

Before we merge, can you revert these?

**Review by marcospb19 — CHANGES_REQUESTED** [review]

File: Cargo.toml
Comment (lmkra) [reply Cargo.toml]: Oops, must have slipped unnoticed during the rebase.
Should be ok now.

**Pull request merged**
